File: web/frontend/lib/core/auth_client.dart
import 'dart:convert';
import 'dart:html' as html;
import 'package:flutter/foundation.dart' show ChangeNotifier, kDebugMode, debugPrint;

const _tokenKey = 'trinity_auth_token';
const _roleKey = 'trinity_auth_role';
const _permissionsKey = 'trinity_auth_permissions';
const _userIdKey = 'trinity_auth_user_id';
const _emailKey = 'trinity_auth_email';
const _activeOpenClawIdKey = 'trinity_active_openclaw_id';

enum AuthRole { guest, user, admin, superadmin }

AuthRole parseRole(String? name) {
  switch (name) {
    case 'superadmin':
      return AuthRole.superadmin;
    case 'admin':
      return AuthRole.admin;
    case 'user':
      return AuthRole.user;
    default:
      return AuthRole.guest;
  }
}

String roleToString(AuthRole role) {
  switch (role) {
    case AuthRole.superadmin:
      return 'superadmin';
    case AuthRole.admin:
      return 'admin';
    case AuthRole.user:
      return 'user';
    case AuthRole.guest:
      return 'guest';
  }
}

class AuthState {
  final String? token;
  final String? userId;
  final String? email;
  final AuthRole role;
  final List<String> permissions;
  final bool isGuest;
  final List<OpenClawInfo> openclaws;
  final String? activeOpenClawId;

  const AuthState({
    this.token,
    this.userId,
    this.email,
    this.role = AuthRole.guest,
    this.permissions = const [],
    this.isGuest = true,
    this.openclaws = const [],
    this.activeOpenClawId,
  });

  bool hasPermission(String action) => permissions.contains(action);

  bool get isAuthenticated => !isGuest && token != null;

  /// The currently selected OpenClaw instance, or null if none selected.
  OpenClawInfo? get activeOpenClaw {
    if (activeOpenClawId == null) return null;
    try {
      return openclaws.firstWhere((oc) => oc.id == activeOpenClawId);
    } catch (_) {
      return null;
    }
  }

  AuthState copyWith({
    String? token,
    String? userId,
    String? email,
    AuthRole? role,
    List<String>? permissions,
    bool? isGuest,
    List<OpenClawInfo>? openclaws,
    String? activeOpenClawId,
    bool clearActiveOpenClawId = false,
  }) {
    return AuthState(
      token: token ?? this.token,
      userId: userId ?? this.userId,
      email: email ?? this.email,
      role: role ?? this.role,
      permissions: permissions ?? this.permissions,
      isGuest: isGuest ?? this.isGuest,
      openclaws: openclaws ?? this.openclaws,
      activeOpenClawId: clearActiveOpenClawId ? null : (activeOpenClawId ?? this.activeOpenClawId),
    );
  }
}

/// Status of the user's assigned OpenClaw instances.
enum OpenClawStatus { unknown, loading, ready, noOpenClaws, error }

/// Describes an admin-managed shared OpenClaw instance assigned to a user.
class OpenClawInfo {
  final String id;
  final String name;
  final String? description;
  final String status;
  final bool ready;
  final int? userCount;

  const OpenClawInfo({
    required this.id,
    required this.name,
    this.description,
    required this.status,
    required this.ready,
    this.userCount,
  });

  factory OpenClawInfo.fromJson(Map<String, dynamic> json) {
    return OpenClawInfo(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      status: json['status'] as String? ?? 'unknown',
      ready: json['ready'] as bool? ?? false,
      userCount: json['userCount'] as int?,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    if (description != null) 'description': description,
    'status': status,
    'ready': ready,
  };
}

class AuthClient extends ChangeNotifier {
  AuthState _state = const AuthState();
  AuthState get state => _state;

  final String authServiceBaseUrl;

  /// Status of the user's assigned OpenClaw instances.
  OpenClawStatus _openClawStatus = OpenClawStatus.unknown;
  OpenClawStatus get openClawStatus => _openClawStatus;

  /// Human-readable error when [openClawStatus] is [OpenClawStatus.error].
  String? _openClawError;
  String? get openClawError => _openClawError;

  AuthClient({required this.authServiceBaseUrl}) {
    _restoreFromStorage();
  }

  void _restoreFromStorage() {
    final token = html.window.localStorage[_tokenKey];
    final role = html.window.localStorage[_roleKey];
    final permsJson = html.window.localStorage[_permissionsKey];
    final userId = html.window.localStorage[_userIdKey];
    final email = html.window.localStorage[_emailKey];
    final activeOpenClawId = html.window.localStorage[_activeOpenClawIdKey];

    if (token != null && token.isNotEmpty) {
      List<String> permissions = [];
      if (permsJson != null && permsJson.isNotEmpty) {
        try {
          permissions = List<String>.from(jsonDecode(permsJson));
        } catch (e) {
          if (kDebugMode) debugPrint('[Auth] Failed to parse stored permissions: $e');
        }
      }
      _state = AuthState(
        token: token,
        userId: userId,
        email: email,
        role: parseRole(role),
        permissions: permissions,
        isGuest: role == 'guest',
        activeOpenClawId: activeOpenClawId,
      );
      notifyListeners();

      // Restored session — fetch the user's assigned OpenClaw instances.
      fetchUserOpenClaws();
    }
  }

  void _persistToStorage() {
    if (_state.token != null) {
      html.window.localStorage[_tokenKey] = _state.token!;
    } else {
      html.window.localStorage.remove(_tokenKey);
    }
    html.window.localStorage[_roleKey] = roleToString(_state.role);
    html.window.localStorage[_permissionsKey] = jsonEncode(_state.permissions);
    if (_state.userId != null) {
      html.window.localStorage[_userIdKey] = _state.userId!;
    }
    if (_state.email != null) {
      html.window.localStorage[_emailKey] = _state.email!;
    }
    if (_state.activeOpenClawId != null) {
      html.window.localStorage[_activeOpenClawIdKey] = _state.activeOpenClawId!;
    } else {
      html.window.localStorage.remove(_activeOpenClawIdKey);
    }
  }

  Future<void> loginWithEmail(String email, String password) async {
    // Call GoTrue signup/signin
    final uri = Uri.parse('$authServiceBaseUrl/supabase/auth/token?grant_type=password');
    final request = html.HttpRequest();
    request.open('POST', uri.toString());
    request.setRequestHeader('Content-Type', 'application/json');
    request.setRequestHeader('apikey', const String.fromEnvironment('SUPABASE_ANON_KEY', defaultValue: ''));

    final completer = _createRequestCompleter(request);
    request.send(jsonEncode({'email': email, 'password': password}));

    final response = await completer;
    final body = jsonDecode(response);
    final accessToken = body['access_token'] as String;

    await _resolveSession(accessToken, email: email);
  }

  Future<void> signUpWithEmail(String email, String password) async {
    final uri = Uri.parse('$authServiceBaseUrl/supabase/auth/signup');
    final request = html.HttpRequest();
    request.open('POST', uri.toString());
    request.setRequestHeader('Content-Type', 'application/json');
    request.setRequestHeader('apikey', const String.fromEnvironment('SUPABASE_ANON_KEY', defaultValue: ''));

    final completer = _createRequestCompleter(request);
    request.send(jsonEncode({'email': email, 'password': password}));

    final response = await completer;
    final body = jsonDecode(response);
    final accessToken = body['access_token'] as String?;

    if (accessToken != null) {
      await _resolveSession(accessToken, email: email);
    }
  }

  Future<void> loginAsGuest() async {
    final uri = Uri.parse('$authServiceBaseUrl/auth/guest');
    final request = html.HttpRequest();
    request.open('POST', uri.toString());
    request.setRequestHeader('Content-Type', 'application/json');

    final completer = _createRequestCompleter(request);
    request.send('{}');

    final response = await completer;
    final body = jsonDecode(response);

    _state = AuthState(
      token: body['token'] as String,
      role: AuthRole.guest,
      permissions: List<String>.from(body['permissions'] ?? []),
      isGuest: true,
    );
    _persistToStorage();
    notifyListeners();

    // Fetch the user's assigned OpenClaw instances.
    fetchUserOpenClaws();
  }

  Future<void> _resolveSession(String accessToken, {String? email}) async {
    final uri = Uri.parse('$authServiceBaseUrl/auth/me');
    final request = html.HttpRequest();
    request.open('GET', uri.toString());
    request.setRequestHeader('Authorization', 'Bearer $accessToken');

    final completer = _createRequestCompleter(request);
    request.send();

    final response = await completer;
    final body = jsonDecode(response);

    _state = AuthState(
      token: accessToken,
      userId: body['id'] as String?,
      email: email ?? body['email'] as String?,
      role: parseRole(body['role'] as String?),
      permissions: List<String>.from(body['permissions'] ?? []),
      isGuest: body['isGuest'] == true,
    );
    _persistToStorage();
    notifyListeners();

    // Fetch the user's assigned OpenClaw instances.
    // Fire-and-forget: the UI observes openClawStatus reactively.
    fetchUserOpenClaws();
  }

  /// Resolve session from an SSO access token (used after OAuth callback).
  Future<void> resolveSessionFromToken(String accessToken) async {
    await _resolveSession(accessToken);
  }

  // ---------------------------------------------------------------------------
  // OpenClaw instance management
  // ---------------------------------------------------------------------------

  /// Fetch the list of shared OpenClaw instances assigned to the current user
  /// via `GET /auth/openclaws`. Called automatically after login or session
  /// restore. The UI can observe [openClawStatus] reactively.
  Future<void> fetchUserOpenClaws() async {
    if (_state.token == null) return;

    _openClawStatus = OpenClawStatus.loading;
    _openClawError = null;
    notifyListeners();

    try {
      final uri = Uri.parse('$authServiceBaseUrl/auth/openclaws');
      final request = html.HttpRequest();
      request.open('GET', uri.toString());
      request.setRequestHeader('Authorization', 'Bearer ${_state.token}');

      final completer = _createRequestCompleter(request);
      request.send();

      final responseText = await completer.timeout(
        const Duration(seconds: 15),
        onTimeout: () {
          throw Exception('Request timed out after 15 seconds');
        },
      );
      final decoded = jsonDecode(responseText);
      // The endpoint returns either a plain array or {"openclaws": [...]}
      final List rawList;
      if (decoded is List) {
        rawList = decoded;
      } else if (decoded is Map) {
        rawList = (decoded['openclaws'] as List?) ?? [];
      } else {
        rawList = [];
      }
      final openclaws = rawList
          .map((e) => OpenClawInfo.fromJson(Map<String, dynamic>.from(e as Map)))
          .toList();

      if (openclaws.isEmpty) {
        _state = _state.copyWith(
          openclaws: openclaws,
          clearActiveOpenClawId: true,
        );
        _openClawStatus = OpenClawStatus.noOpenClaws;
        _persistToStorage();
        notifyListeners();
        if (kDebugMode) debugPrint('[Auth] No OpenClaws assigned to user');
        return;
      }

      // Auto-select: prefer the previously persisted ID if it's still in the
      // list, otherwise pick the first ready instance, or just the first one.
      final persistedId = _state.activeOpenClawId;
      String selectedId;
      if (persistedId != null && openclaws.any((oc) => oc.id == persistedId)) {
        selectedId = persistedId;
      } else {
        final firstReady = openclaws.where((oc) => oc.ready).toList();
        selectedId = firstReady.isNotEmpty ? firstReady.first.id : openclaws.first.id;
      }

      _state = _state.copyWith(
        openclaws: openclaws,
        activeOpenClawId: selectedId,
      );
      _openClawStatus = OpenClawStatus.ready;
      _persistToStorage();
      notifyListeners();
      if (kDebugMode) {
        debugPrint('[Auth] Fetched ${openclaws.length} OpenClaw(s), active: $selectedId');
      }
    } catch (e) {
      _openClawStatus = OpenClawStatus.error;
      _openClawError = 'Failed to fetch OpenClaws: $e';
      notifyListeners();
      if (kDebugMode) debugPrint('[Auth] fetchUserOpenClaws error: $e');
    }
  }

  /// Switch the active OpenClaw instance. The [id] must be present in the
  /// current [state.openclaws] list.
  void selectOpenClaw(String id) {
    if (!_state.openclaws.any((oc) => oc.id == id)) {
      if (kDebugMode) debugPrint('[Auth] selectOpenClaw: unknown id $id');
      return;
    }
    _state = _state.copyWith(activeOpenClawId: id);
    _persistToStorage();
    notifyListeners();
    if (kDebugMode) debugPrint('[Auth] Active OpenClaw switched to $id');
  }

  void logout() {
    _state = const AuthState();
    _openClawStatus = OpenClawStatus.unknown;
    _openClawError = null;
    html.window.localStorage.remove(_tokenKey);
    html.window.localStorage.remove(_roleKey);
    html.window.localStorage.remove(_permissionsKey);
    html.window.localStorage.remove(_userIdKey);
    html.window.localStorage.remove(_emailKey);
    html.window.localStorage.remove(_activeOpenClawIdKey);
    notifyListeners();
  }

  /// Fetch all users with roles. Requires [users.list] permission.
  Future<List<Map<String, dynamic>>> fetchUsers() async {
    final uri = Uri.parse('$authServiceBaseUrl/auth/users');
    final request = html.HttpRequest();
    request.open('GET', uri.toString());
    request.setRequestHeader('Authorization', 'Bearer ${_state.token}');

    final completer = _createRequestCompleter(request);
    request.send();

    final response = await completer;
    final body = jsonDecode(response);
    return List<Map<String, dynamic>>.from(
      (body['users'] as List?)?.map((e) => Map<String, dynamic>.from(e as Map)) ?? [],
    );
  }

  /// Assign a role to a user. Requires [users.manage] permission.
  Future<void> assignUserRole(String userId, String role) async {
    final uri = Uri.parse('$authServiceBaseUrl/auth/users/$userId/role');
    final request = html.HttpRequest();
    request.open('POST', uri.toString());
    request.setRequestHeader('Content-Type', 'application/json');
    request.setRequestHeader('Authorization', 'Bearer ${_state.token}');

    final completer = _createRequestCompleter(request);
    request.send(jsonEncode({'role': role}));

    await completer;
  }

  /// Fetch paginated audit log. Requires [audit.read] permission.
  Future<List<Map<String, dynamic>>> fetchAuditLog({int limit = 50, int offset = 0}) async {
    final uri = Uri.parse('$authServiceBaseUrl/auth/users/audit?limit=$limit&offset=$offset');
    final request = html.HttpRequest();
    request.open('GET', uri.toString());
    request.setRequestHeader('Authorization', 'Bearer ${_state.token}');

    final completer = _createRequestCompleter(request);
    request.send();

    final response = await completer;
    final body = jsonDecode(response);
    return List<Map<String, dynamic>>.from(
      (body['logs'] as List?)?.map((e) => Map<String, dynamic>.from(e as Map)) ?? [],
    );
  }

  /// Fetch role-permission matrix. Requires [users.list] permission.
  Future<Map<String, dynamic>> fetchRolePermissionMatrix() async {
    final uri = Uri.parse('$authServiceBaseUrl/auth/users/roles/permissions');
    final request = html.HttpRequest();
    request.open('GET', uri.toString());
    request.setRequestHeader('Authorization', 'Bearer ${_state.token}');

    final completer = _createRequestCompleter(request);
    request.send();

    final response = await completer;
    return Map<String, dynamic>.from(jsonDecode(response) as Map);
  }

  /// Update permissions for a role. Requires [users.manage] permission.
  Future<void> updateRolePermissions(String role, List<String> permissions) async {
    final uri = Uri.parse('$authServiceBaseUrl/auth/users/roles/$role/permissions');
    final request = html.HttpRequest();
    request.open('PUT', uri.toString());
    request.setRequestHeader('Content-Type', 'application/json');
    request.setRequestHeader('Authorization', 'Bearer ${_state.token}');

    final completer = _createRequestCompleter(request);
    request.send(jsonEncode({'permissions': permissions}));

    await completer;
  }

  String getKeycloakLoginUrl() {
    return '$authServiceBaseUrl/supabase/auth/authorize?provider=keycloak';
  }

  Future<String> _createRequestCompleter(html.HttpRequest request) {
    final completer = Future<String>.delayed(Duration.zero, () async {
      await request.onLoadEnd.first;
      if (request.status! >= 200 && request.status! < 300) {
        return request.responseText ?? '{}';
      }
      throw Exception('HTTP ${request.status}: ${request.responseText}');
    });
    return completer;
  }
}

File: app/web/src/lib/clients/auth-client.ts
/**
 * AuthClient — 1:1 port of core/auth_client.dart
 *
 * Manages JWT authentication, login flows, OpenClaw instance management,
 * and admin operations. Persists state to localStorage.
 */

/* ------------------------------------------------------------------ */
/*  Types                                                              */
/* ------------------------------------------------------------------ */

export type AuthRole = 'guest' | 'user' | 'admin' | 'superadmin';

export interface OpenClawInfo {
  id: string;
  name: string;
  description?: string;
  status: string;
  ready: boolean;
  userCount: number;
}

export type OpenClawStatus = 'unknown' | 'loading' | 'ready' | 'noOpenClaws' | 'error';

export interface AuthState {
  token: string | null;
  userId: string | null;
  email: string | null;
  role: AuthRole;
  permissions: string[];
  isGuest: boolean;
  openclaws: OpenClawInfo[];
  activeOpenClawId: string | null;
  openClawStatus: OpenClawStatus;
}

export interface AuditLogResult {
  logs: any[];
  total: number;
}

/* ------------------------------------------------------------------ */
/*  Helpers                                                            */
/* ------------------------------------------------------------------ */

function parseRole(s?: string | null): AuthRole {
  switch (s) {
    case 'superadmin':
      return 'superadmin';
    case 'admin':
      return 'admin';
    case 'user':
      return 'user';
    default:
      return 'guest';
  }
}

function parseJwtExpiry(token: string): Date | null {
  try {
    const parts = token.split('.');
    if (parts.length < 2) return null;
    const payload = JSON.parse(atob(parts[1].replace(/-/g, '+').replace(/_/g, '/')));
    if (typeof payload.exp === 'number') {
      return new Date(payload.exp * 1000);
    }
    return null;
  } catch {
    return null;
  }
}

function isExpiredToken(token: string): boolean {
  const exp = parseJwtExpiry(token);
  if (!exp) return false;
  return exp.getTime() < Date.now();
}

const STORAGE_PREFIX = 'trinity_auth_';

function storageGet(key: string): string | null {
  try {
    return localStorage.getItem(STORAGE_PREFIX + key);
  } catch {
    return null;
  }
}

function storageSet(key: string, value: string): void {
  try {
    localStorage.setItem(STORAGE_PREFIX + key, value);
  } catch {
    // Ignore storage errors
  }
}

function storageRemove(key: string): void {
  try {
    localStorage.removeItem(STORAGE_PREFIX + key);
  } catch {
    // Ignore storage errors
  }
}

/* ------------------------------------------------------------------ */
/*  AuthClient                                                         */
/* ------------------------------------------------------------------ */

export type AuthListener = (state: AuthState) => void;

export class AuthClient {
  private _state: AuthState;
  private _baseUrl: string;
  private _listeners = new Set<AuthListener>();
  private _openClawFetchSeq = 0;

  constructor(baseUrl: string) {
    this._baseUrl = baseUrl;
    this._state = {
      token: null,
      userId: null,
      email: null,
      role: 'guest',
      permissions: [],
      isGuest: true,
      openclaws: [],
      activeOpenClawId: null,
      openClawStatus: 'unknown',
    };
    this._restoreFromStorage();
  }

  get state(): AuthState {
    return this._state;
  }

  subscribe(fn: AuthListener): () => void {
    this._listeners.add(fn);
    return () => this._listeners.delete(fn);
  }

  private _notify(): void {
    this._listeners.forEach((fn) => fn(this._state));
  }

  private _setState(patch: Partial<AuthState>): void {
    this._state = { ...this._state, ...patch };
    this._notify();
  }

  /* ---------------------------------------------------------------- */
  /*  Storage                                                          */
  /* ---------------------------------------------------------------- */

  private _restoreFromStorage(): void {
    const token = storageGet('token');
    if (!token) return;
    if (isExpiredToken(token)) {
      this.logout();
      return;
    }

    const role = parseRole(storageGet('role'));
    const permissions = (() => {
      try {
        const raw = storageGet('permissions');
        return raw ? JSON.parse(raw) : [];
      } catch {
        return [];
      }
    })();

    this._state = {
      token,
      userId: storageGet('userId'),
      email: storageGet('email'),
      role,
      permissions,
      isGuest: role === 'guest',
      openclaws: [],
      activeOpenClawId: storageGet('activeOpenClawId'),
      openClawStatus: 'loading',
    };

    // Fetch openclaws in background
    if (role !== 'guest') {
      this.fetchUserOpenClaws();
    }
  }

  private _persistState(): void {
    const s = this._state;
    if (s.token) {
      storageSet('token', s.token);
      storageSet('role', s.role);
      storageSet('permissions', JSON.stringify(s.permissions));
      if (s.userId) storageSet('userId', s.userId);
      if (s.email) storageSet('email', s.email);
      if (s.activeOpenClawId) storageSet('activeOpenClawId', s.activeOpenClawId);
    }
  }

  private _clearStorage(): void {
    const keys = ['token', 'role', 'permissions', 'userId', 'email', 'activeOpenClawId'];
    keys.forEach((k) => storageRemove(k));
  }

  /* ---------------------------------------------------------------- */
  /*  Login flows                                                      */
  /* ---------------------------------------------------------------- */

  async loginWithEmail(email: string, password: string): Promise<void> {
    const res = await fetch(`${this._baseUrl}/supabase/auth/token?grant_type=password`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email, password }),
    });
    if (!res.ok) {
      const body = await res.json().catch(() => ({}));
      throw new Error(body.msg ?? body.error_description ?? 'Login failed');
    }
    const data = await res.json();
    await this._resolveSession(data.access_token);
  }

  async signUpWithEmail(email: string, password: string): Promise<void> {
    const res = await fetch(`${this._baseUrl}/supabase/auth/signup`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email, password }),
    });
    if (!res.ok) {
      const body = await res.json().catch(() => ({}));
      throw new Error(body.msg ?? body.error_description ?? 'Sign up failed');
    }
  }

  async loginAsGuest(): Promise<void> {
    const res = await fetch(`${this._baseUrl}/auth/guest`, { method: 'POST' });
    if (!res.ok) {
      throw new Error('Guest login failed');
    }
    const data = await res.json();
    await this._resolveSession(data.token ?? data.access_token);
  }

  async resolveSessionFromToken(accessToken: string): Promise<void> {
    await this._resolveSession(accessToken);
  }

  private async _resolveSession(accessToken: string): Promise<void> {
    const ts = Date.now();
    const res = await fetch(`${this._baseUrl}/auth/me?ts=${ts}`, {
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Cache-Control': 'no-cache',
      },
    });
    if (!res.ok) throw new Error('Failed to resolve session');
    const me = await res.json();

    const role = parseRole(me.role);
    this._setState({
      token: accessToken,
      userId: me.id ?? me.userId,
      email: me.email,
      role,
      permissions: me.permissions ?? [],
      isGuest: me.isGuest === true || role === 'guest',
      openClawStatus: role === 'guest' ? 'ready' : 'loading',
    });
    this._persistState();

    if (role !== 'guest') {
      await this.fetchUserOpenClaws();
    }
  }

  logout(): void {
    this._clearStorage();
    this._state = {
      token: null,
      userId: null,
      email: null,
      role: 'guest',
      permissions: [],
      isGuest: true,
      openclaws: [],
      activeOpenClawId: null,
      openClawStatus: 'unknown',
    };
    this._notify();
  }

  hasPermission(action: string): boolean {
    return this._state.permissions.includes(action);
  }

  /* ---------------------------------------------------------------- */
  /*  OpenClaw management                                              */
  /* ---------------------------------------------------------------- */

  async fetchUserOpenClaws(): Promise<void> {
    const seq = ++this._openClawFetchSeq;
    const token = this._state.token;
    if (!token) return;

    this._setState({ openClawStatus: 'loading' });

    try {
      const res = await fetch(`${this._baseUrl}/auth/openclaws`, {
        headers: { Authorization: `Bearer ${token}` },
      });
      if (!res.ok) throw new Error('Failed to fetch openclaws');
      if (seq !== this._openClawFetchSeq) return; // stale

      const data = await res.json();
      const list: OpenClawInfo[] = (Array.isArray(data) ? data : data.openclaws ?? []).map(
        (o: any) => ({
          id: o.id,
          name: o.name ?? o.id,
          description: o.description,
          status: o.status ?? 'unknown',
          ready: o.ready === true,
          userCount: o.userCount ?? o.user_count ?? 0,
        }),
      );

      const savedId = this._state.activeOpenClawId;
      const activeId =
        (savedId && list.find((o) => o.id === savedId)?.id) ??
        list.find((o) => o.ready)?.id ??
        list[0]?.id ??
        null;

      this._setState({
        openclaws: list,
        activeOpenClawId: activeId,
        openClawStatus: list.length === 0 ? 'noOpenClaws' : 'ready',
      });
      if (activeId) storageSet('activeOpenClawId', activeId);
    } catch (err) {
      if (seq !== this._openClawFetchSeq) return;
      // Retry once after 300ms
      await new Promise((r) => setTimeout(r, 300));
      if (seq !== this._openClawFetchSeq) return;

      try {
        const res2 = await fetch(`${this._baseUrl}/auth/openclaws`, {
          headers: { Authorization: `Bearer ${token}` },
        });
        if (res2.ok && seq === this._openClawFetchSeq) {
          const data = await res2.json();
          const list: OpenClawInfo[] = (Array.isArray(data) ? data : data.openclaws ?? []).map(
            (o: any) => ({
              id: o.id,
              name: o.name ?? o.id,
              description: o.description,
              status: o.status ?? 'unknown',
              ready: o.ready === true,
              userCount: o.userCount ?? o.user_count ?? 0,
            }),
          );
          const activeId = list.find((o) => o.ready)?.id ?? list[0]?.id ?? null;
          this._setState({
            openclaws: list,
            activeOpenClawId: activeId,
            openClawStatus: list.length === 0 ? 'noOpenClaws' : 'ready',
          });
        }
      } catch {
        if (seq === this._openClawFetchSeq) {
          this._setState({ openClawStatus: 'error' });
        }
      }
    }
  }

  selectOpenClaw(id: string): void {
    const found = this._state.openclaws.find((o) => o.id === id);
    if (!found) return;
    this._setState({ activeOpenClawId: id });
    storageSet('activeOpenClawId', id);
  }

  get activeOpenClaw(): OpenClawInfo | undefined {
    return this._state.openclaws.find((o) => o.id === this._state.activeOpenClawId);
  }

  /* ---------------------------------------------------------------- */
  /*  Admin APIs                                                       */
  /* ---------------------------------------------------------------- */

  private async _authedFetch(path: string, options: RequestInit = {}): Promise<Response> {
    return fetch(`${this._baseUrl}${path}`, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this._state.token}`,
        ...(options.headers ?? {}),
      },
    });
  }

  async fetchUsers(): Promise<any[]> {
    const res = await this._authedFetch('/auth/users');
    if (!res.ok) throw new Error('Failed to fetch users');
    return res.json();
  }

  async assignUserRole(userId: string, role: string): Promise<void> {
    const res = await this._authedFetch(`/auth/users/${userId}/role`, {
      method: 'POST',
      body: JSON.stringify({ role }),
    });
    if (!res.ok) throw new Error('Failed to assign role');
  }

  async fetchAuditLog(params: {
    limit?: number;
    offset?: number;
    action?: string;
    userId?: string;
    resource?: string;
    ip?: string;
    from?: string;
    to?: string;
  }): Promise<AuditLogResult> {
    const qs = new URLSearchParams();
    if (params.limit) qs.set('limit', String(params.limit));
    if (params.offset) qs.set('offset', String(params.offset));
    if (params.action) qs.set('action', params.action);
    if (params.userId) qs.set('user_id', params.userId);
    if (params.resource) qs.set('resource', params.resource);
    if (params.ip) qs.set('ip', params.ip);
    if (params.from) qs.set('from', params.from);
    if (params.to) qs.set('to', params.to);

    const res = await this._authedFetch(`/auth/users/audit?${qs}`);
    if (!res.ok) throw new Error('Failed to fetch audit log');
    const data = await res.json();
    return { logs: data.logs ?? data, total: data.total ?? 0 };
  }

  getAuditExportUrl(format: 'csv' | 'json', params: Record<string, string> = {}): string {
    const qs = new URLSearchParams({ format, ...params });
    return `${this._baseUrl}/auth/users/audit/export?${qs}`;
  }

  async fetchRolePermissionMatrix(): Promise<Record<string, string[]>> {
    const res = await this._authedFetch('/auth/users/roles/permissions');
    if (!res.ok) throw new Error('Failed to fetch role permissions');
    return res.json();
  }

  async updateRolePermissions(role: string, permissions: string[]): Promise<void> {
    const res = await this._authedFetch(`/auth/users/roles/${role}/permissions`, {
      method: 'PUT',
      body: JSON.stringify({ permissions }),
    });
    if (!res.ok) throw new Error('Failed to update role permissions');
  }

  getKeycloakLoginUrl(): string {
    return `${this._baseUrl}/supabase/auth/authorize?provider=keycloak`;
  }
}
